src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs

```diff
@@ -187,48 +261,52 @@ private Type0CommandMessageSection<BsonDocument> CreateType0Section(ConnectionDe
-            extraElements.Add(dbElement);
+            addIfNotAlreadyAdded("$db", _databaseNamespace.DatabaseName);
 
             if (_readPreference != null && _readPreference != ReadPreference.Primary)
             {
                 var readPreferenceDocument = QueryHelper.CreateReadPreferenceDocument(_readPreference);
-                var readPreferenceElement = new BsonElement("$readPreference", readPreferenceDocument);
-                extraElements.Add(readPreferenceElement);
+                addIfNotAlreadyAdded("$readPreference", readPreferenceDocument);
             }
 
             if (_session.Id != null)
             {
-                var lsidElement = new BsonElement("lsid", _session.Id);
-                extraElements.Add(lsidElement);
+                addIfNotAlreadyAdded("lsid", _session.Id);
             }
 
             if (_session.ClusterTime != null)
             {
-                var clusterTimeElement = new BsonElement("$clusterTime", _session.ClusterTime);
-                extraElements.Add(clusterTimeElement);
+                addIfNotAlreadyAdded("$clusterTime", _session.ClusterTime);
             }
             Action<BsonWriterSettings> writerSettingsConfigurator = s => s.GuidRepresentation = GuidRepresentation.Unspecified;
 
             _session.AboutToSendCommand();
             if (_session.IsInTransaction)
             {
                 var transaction = _session.CurrentTransaction;
-                extraElements.Add(new BsonElement("txnNumber", transaction.TransactionNumber));
+                addIfNotAlreadyAdded("txnNumber", transaction.TransactionNumber);
                 if (transaction.State == CoreTransactionState.Starting)
                 {
-                    extraElements.Add(new BsonElement("startTransaction", true));
+                    addIfNotAlreadyAdded("startTransaction", true);
                     var readConcern = ReadConcernHelper.GetReadConcernForFirstCommandInTransaction(_session, connectionDescription);
                     if (readConcern != null)
                     {
-                        extraElements.Add(new BsonElement("readConcern", readConcern));
+                        addIfNotAlreadyAdded("readConcern", readConcern);
                     }
                 }
-                extraElements.Add(new BsonElement("autocommit", false));
+                addIfNotAlreadyAdded("autocommit", false);
             }
 
             var elementAppendingSerializer = new ElementAppendingSerializer<BsonDocument>(BsonDocumentSerializer.Instance, extraElements, writerSettingsConfigurator);
             return new Type0CommandMessageSection<BsonDocument>(_command, elementAppendingSerializer);
+
+            void addIfNotAlreadyAdded(string key, object value)
+            {
+                if (!_command.Contains(key))
+                {
+                    extraElements.Add(new BsonElement(key, BsonValue.Create(value)));
```

Comment:
Don't use `BsonValue.Create(object)` overload. We're probably going to deprecate it (it ignores any configured serializers).

After changing the type of the `value` parameter this line can be just:

    extraElements.Add(new BsonElement(key, BsonValue.Create(value)));